### ISSUE TITLE
Fentry failing to load on arm64

### ISSUE THREAD

**Issue by samp66** [libbpf/libbpf-bootstrap#56]
I am successfully building bootstrap, minimal, uprobe examples and they all work on M1 Macbook running an arm64 ubuntu VM but Fentry fails for some odd reason I can't figure out. Any tips on how to debug such issues:

```
$ uname -a
Linux devvm 5.13.0-21-generic #21-Ubuntu SMP Tue Oct 19 09:01:50 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux
```

```
$ sudo ./fentry
[sudo] password for sasan:
libbpf: loading object 'fentry_bpf' from buffer
libbpf: elf: section(2) fentry/do_unlinkat, size 200, link 0, flags 6, type=1
libbpf: sec 'fentry/do_unlinkat': found program 'do_unlinkat' at insn offset 0 (0 bytes), code size 25 insns (200 bytes)
libbpf: elf: section(3) fexit/do_unlinkat, size 256, link 0, flags 6, type=1
libbpf: sec 'fexit/do_unlinkat': found program 'do_unlinkat_exit' at insn offset 0 (0 bytes), code size 32 insns (256 bytes)
libbpf: elf: section(4) license, size 13, link 0, flags 3, type=1
libbpf: license of fentry_bpf is Dual BSD/GPL
libbpf: elf: section(5) .rodata.str1.1, size 76, link 0, flags 32, type=1
libbpf: elf: skipping unrecognized data section(5) .rodata.str1.1
libbpf: elf: section(6) .BTF, size 945, link 0, flags 0, type=1
libbpf: elf: section(7) .BTF.ext, size 364, link 0, flags 0, type=1
libbpf: elf: section(8) .symtab, size 144, link 12, flags 0, type=2
libbpf: looking for externs among 6 symbols...
libbpf: collected 0 externs total
libbpf: loading kernel BTF '/sys/kernel/btf/vmlinux': 0
libbpf: sec 'fentry/do_unlinkat': found 1 CO-RE relocations
libbpf: prog 'do_unlinkat': relo #0: kind <byte_off> (0), spec is [6] struct filename.name (0:0 @ offset 0)
libbpf: CO-RE relocating [0] struct filename: found target candidate [1115] struct filename in [vmlinux]
libbpf: prog 'do_unlinkat': relo #0: matching candidate #0 [1115] struct filename.name (0:0 @ offset 0)
libbpf: prog 'do_unlinkat': relo #0: patched insn #16 (LDX/ST/STX) off 0 -> 0
libbpf: sec 'fexit/do_unlinkat': found 1 CO-RE relocations
libbpf: prog 'do_unlinkat_exit': relo #0: kind <byte_off> (0), spec is [6] struct filename.name (0:0 @ offset 0)
libbpf: prog 'do_unlinkat_exit': relo #0: matching candidate #0 [1115] struct filename.name (0:0 @ offset 0)
libbpf: prog 'do_unlinkat_exit': relo #0: patched insn #22 (LDX/ST/STX) off 0 -> 0
libbpf: prog 'do_unlinkat': failed to attach: ERROR: strerror_r(-524)=22
libbpf: failed to auto-attach program 'do_unlinkat': -524
Failed to attach BPF skeleton
```

**Comment by anakryiko** [comment]
fentry is only supported on x86 architecture, unfortunately. arch_prepare_bpf_trampoline() has to be implemented for arm64 to allow this (at the very least).

**Comment by samp66** [comment]
Is there an easy way to check which calls are supported on arm64 vs x86? I looked around for quite a bit and wasn't able to tell if this was a config issue vs a support problem.

**Comment by anakryiko** [comment]
I don't think there is any easy way to check that except trying to load a simplistic fentry program, sorry.

**Comment by Jeftah** [comment]
If I see it correctly, then with https://github.com/torvalds/linux/commit/efc9909fdce00a827a37609628223cd45bf95d0b fentry should be available for arm64 since kernel version 6.0.

Unfortunately I have not been able to get it to work on the Raspberry Pi 4 B with kernel version 6.0 and 6.1RC8 fentry and get the same error message as @samp66.

bootstrap, kprobe, minimal_legacy, minimal, sockfilter, uprobe and tc work.

Is there any factor I have not considered?

**Comment by anakryiko** [comment]
@Jeftah  I can only recommend trying https://github.com/anakryiko/retsnoop or some similar tool to understand where the failure in kernel originates from, and go from there.
